src/ptrace.cc

```diff
@@ -93,4 +150,30 @@ std::unique_ptr<uint8_t[]> PtracePeekBytes(pid_t pid, unsigned long addr,
   }
   return bytes;
 }
+
+#ifdef __amd64__
+long PtraceCallFunction(pid_t pid, unsigned long addr) {
```

Comment:
I believe this code is incorrect in certain corner cases. Notably, there's a problem if `%rip` is at or near `addr`. For instance, let's say you want to call some function `foo()`. If the process is already in `foo()` when you ptrace attach, then when you insert your new code at `%rip` the subsequent call to `foo()` will fail: it could deadlock if `foo()` is guarded by a mutex, or it could become an infinite loop.

This is a problem I had using the same technique on another project at Uber (unfortunately not open source), and why I didn't implement code execution earlier in Pyflame. In that case I was running into this issue a lot because I was trying to call common libc functions like `malloc()`.

I believe the correct way to do this is:
 * use PTRACE_POKETEXT to copy the bytes to an unused portion of memory
 * use PTRACE_SETREGS to point the program counter to the memory area, and resume the process

One way to do this is to directly make a syscall to `mmap(2)` to get a single page of memory, and then use that page. There's also usually room for a few extra bytes of code at the end of the `.text` area, but that's not guaranteed.

What do you think?